diff --git a/dist/declarations/src/actors/promise.d.ts b/dist/declarations/src/actors/promise.d.ts
index 350bb3373e502d6f54aa1fb2dafcaaf208d57a8c..706823d8676823106d24ae2cff51b4784e8e5077 100644
--- a/dist/declarations/src/actors/promise.d.ts
+++ b/dist/declarations/src/actors/promise.d.ts
@@ -1,9 +1,9 @@
 import { AnyActorSystem } from "../system.js";
 import { ActorLogic, ActorRefFromLogic, EventObject, NonReducibleUnknown, Snapshot } from "../types.js";
-export type PromiseSnapshot<TOutput, TInput> = Snapshot<TOutput> & {
+export type PromiseSnapshot<TOutput, TInput, TError = unknown> = Snapshot<TOutput, TError> & {
     input: TInput | undefined;
 };
-export type PromiseActorLogic<TOutput, TInput = unknown, TEmitted extends EventObject = EventObject> = ActorLogic<PromiseSnapshot<TOutput, TInput>, {
+export type PromiseActorLogic<TOutput, TInput = unknown, TEmitted extends EventObject = EventObject, TError = unknown> = ActorLogic<PromiseSnapshot<TOutput, TInput, TError>, {
     type: string;
     [k: string]: unknown;
 }, TInput, // input
@@ -41,7 +41,7 @@ AnyActorSystem, TEmitted>;
  *
  * @see {@link fromPromise}
  */
-export type PromiseActorRef<TOutput> = ActorRefFromLogic<PromiseActorLogic<TOutput, unknown>>;
+export type PromiseActorRef<TOutput, TError = unknown> = ActorRefFromLogic<PromiseActorLogic<TOutput, unknown, EventObject, TError>>;
 /**
  * An actor logic creator which returns promise logic as defined by an async
  * process that resolves or rejects after some time.
@@ -92,13 +92,13 @@ export type PromiseActorRef<TOutput> = ActorRefFromLogic<PromiseActorLogic<TOutp
  *
  * @see {@link https://stately.ai/docs/input | Input docs} for more information about how input is passed
  */
-export declare function fromPromise<TOutput, TInput = NonReducibleUnknown, TEmitted extends EventObject = EventObject>(promiseCreator: ({ input, system, self, signal, emit }: {
+export declare function fromPromise<TOutput, TInput = NonReducibleUnknown, TEmitted extends EventObject = EventObject, TError = unknown>(promiseCreator: ({ input, system, self, signal, emit }: {
     /** Data that was provided to the promise actor */
     input: TInput;
     /** The actor system to which the promise actor belongs */
     system: AnyActorSystem;
     /** The parent actor of the promise actor */
-    self: PromiseActorRef<TOutput>;
+    self: PromiseActorRef<TOutput, TError>;
     signal: AbortSignal;
     emit: (emitted: TEmitted) => void;
-}) => PromiseLike<TOutput>): PromiseActorLogic<TOutput, TInput, TEmitted>;
+}) => PromiseLike<TOutput>): PromiseActorLogic<TOutput, TInput, TEmitted, TError>;
diff --git a/dist/declarations/src/types.d.ts b/dist/declarations/src/types.d.ts
index a9b962a21b0b3deacd9ff70c4e1e05906edf2794..4080fd36567471801d804802fe96810a33cb31f8 100644
--- a/dist/declarations/src/types.d.ts
+++ b/dist/declarations/src/types.d.ts
@@ -96,6 +96,11 @@ export type OutputFrom<T> = T extends ActorLogic<infer TSnapshot, infer _TEvent,
 })['output'] : T extends ActorRef<infer TSnapshot, infer _TEvent, infer _TEmitted> ? (TSnapshot & {
     status: 'done';
 })['output'] : never;
+export type ErrorFrom<T> = T extends ActorLogic<infer TSnapshot, infer _TEvent, infer _TInput, infer _TSystem, infer _TEmitted> ? (TSnapshot & {
+    status: 'error';
+})['error'] : T extends ActorRef<infer TSnapshot, infer _TEvent, infer _TEmitted> ? (TSnapshot & {
+    status: 'error';
+})['error'] : never;
 export type ActionFunction<TContext extends MachineContext, TExpressionEvent extends EventObject, TEvent extends EventObject, TParams extends ParameterizedObject['params'] | undefined, TActor extends ProvidedActor, TAction extends ParameterizedObject, TGuard extends ParameterizedObject, TDelay extends string, TEmitted extends EventObject> = {
     (args: ActionArgs<TContext, TExpressionEvent, TEvent>, params: TParams): void;
     _out_TEvent?: TEvent;
@@ -218,7 +223,7 @@ type DistributeActors<TContext extends MachineContext, TEvent extends EventObjec
      * The transition to take upon the invoked child machine sending an
      * error event.
      */
-    onError?: string | SingleOrArray<TransitionConfigOrTarget<TContext, ErrorActorEvent, TEvent, TActor, TAction, TGuard, TDelay, TEmitted, TMeta>>;
+    onError?: string | SingleOrArray<TransitionConfigOrTarget<TContext, ErrorActorEvent<ErrorFrom<TSpecificActor['logic']>>, TEvent, TActor, TAction, TGuard, TDelay, TEmitted, TMeta>>;
     onSnapshot?: string | SingleOrArray<TransitionConfigOrTarget<TContext, SnapshotEvent<SnapshotFrom<TSpecificActor['logic']>>, TEvent, TActor, TAction, TGuard, TDelay, TEmitted, TMeta>>;
 } & {
     [K in RequiredActorOptions<TSpecificActor>]: unknown;
@@ -788,7 +793,7 @@ export type AnyActorScope = ActorScope<any, // TSnapshot
 any, // TEvent
 AnyActorSystem, any>;
 export type SnapshotStatus = 'active' | 'done' | 'error' | 'stopped';
-export type Snapshot<TOutput> = {
+export type Snapshot<TOutput, TError = unknown> = {
     status: 'active';
     output: undefined;
     error: undefined;
@@ -799,7 +804,7 @@ export type Snapshot<TOutput> = {
 } | {
     status: 'error';
     output: undefined;
-    error: unknown;
+    error: TError;
 } | {
     status: 'stopped';
     output: undefined;
